daft/src/lib.rs

```diff
@@ -611,48 +628,107 @@ macro_rules! map_diff {
-                pub unchanged: $typ<&'daft K, &'daft V>,
+            pub struct [<$typ Diff>]<'daft, K: $key_constraint + Eq, V> {
+                pub common: $typ<&'daft K, Leaf<'daft, V>>,
                 pub added: $typ<&'daft K, &'daft V>,
                 pub removed: $typ<&'daft K, &'daft V>,
-                pub modified: $typ<&'daft K, V::Diff<'daft>>,
             }
 
-            impl<'daft, K: $key_constraint + Eq, V: Diffable> [<$typ Diff>]<'daft, K, V> {
+            impl<'daft, K: $key_constraint + Eq, V> [<$typ Diff>]<'daft, K, V> {
                 pub fn new() -> Self {
                     Self {
-                        unchanged: $typ::new(),
+                        common: $typ::new(),
                         added: $typ::new(),
                         removed: $typ::new(),
-                        modified: $typ::new(),
                     }
                 }
             }
 
+            impl<'daft, K: $key_constraint + Eq, V: Eq> [<$typ Diff>]<'daft, K, V> {
+                /// Return an iterator over unchanged keys and values.
+                pub fn unchanged(&self) -> impl Iterator<Item = (&'daft K, &'daft V)> + '_ {
+                    self.common.iter().filter_map(|(k, leaf)| {
+                        (leaf.before == leaf.after).then_some((*k, leaf.before))
+                    })
+                }
+
+                /// Return an iterator over unchanged keys.
+                pub fn unchanged_keys(&self) -> impl Iterator<Item = &'daft K> + '_ {
+                    self.common.iter().filter_map(|(k, leaf)| {
+                        (leaf.before == leaf.after).then_some(*k)
+                    })
+                }
+
+                /// Return an iterator over unchanged values.
+                pub fn unchanged_values(&self) -> impl Iterator<Item = &'daft V> + '_ {
+                    self.common.iter().filter_map(|(_, leaf)| {
+                        (leaf.before == leaf.after).then_some(leaf.before)
+                    })
+                }
+
+                /// Return an iterator over modified keys and values.
+                pub fn modified(&self) -> impl Iterator<Item = (&'daft K, Leaf<'daft, V>)> + '_ {
+                    self.common.iter().filter_map(|(k, leaf)| {
+                        (leaf.before != leaf.after).then_some((*k, *leaf))
+                    })
+                }
+
+                /// Return an iterator over modified keys.
+                pub fn modified_keys(&self) -> impl Iterator<Item = &'daft K> + '_ {
+                    self.common.iter().filter_map(|(k, leaf)| {
+                        (leaf.before != leaf.after).then_some(*k)
+                    })
+                }
+
+                /// Return an iterator over modified values.
+                pub fn modified_values(&self) -> impl Iterator<Item = Leaf<'daft, V>> + '_ {
+                    self.common.iter().filter_map(|(_, leaf)| {
+                        (leaf.before != leaf.after).then_some(*leaf)
```

Comment:
TIL: `then_some` is super handy for `filter_map`!

Reply:
Yeah, both `then` (which takes a closure) and `then_some` (which does not) are quite nice.